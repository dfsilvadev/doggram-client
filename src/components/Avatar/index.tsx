import { FC } from "react";
import { Link } from "react-router-dom";
import * as DropdownMenu from "@radix-ui/react-dropdown-menu";
import { SignOut, User } from "phosphor-react";

import { ENDPOINT } from "@/utils/common/constant/endpoints";

import * as S from "./styles";

import { AvatarProps } from "./types";

const Avatar: FC<AvatarProps> = ({ ...props }) => {
  return (
    <DropdownMenu.Root>
      <DropdownMenu.Trigger asChild>
        <S.Image {...props} aria-label="avatar" />
      </DropdownMenu.Trigger>

      <DropdownMenu.Portal>
        <S.DropdownMenuContent sideOffset={5} aria-label="profile menu">
          <S.DropdownMenuItem>
            <Link to={ENDPOINT.PROFILE.HOME}>
              <User weight="bold" /> <span>Perfil</span>
            </Link>
          </S.DropdownMenuItem>

          <S.DropdownMenuSeparator />

          <S.DropdownMenuItem>
            <SignOut weight="bold" /> <span>Sign out</span>
          </S.DropdownMenuItem>

          <S.DropdownMenuArrow />
        </S.DropdownMenuContent>
      </DropdownMenu.Portal>
    </DropdownMenu.Root>
  );
};

export default Avatar;
