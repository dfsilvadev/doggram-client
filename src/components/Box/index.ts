import styled, { css } from "styled-components";

const Box = styled.section`
  ${({ theme }) => css`
    background-color: ${theme.colors.white};
    padding: 1.5rem;
    margin-top: 1.5rem;
  `}
`;

export default Box;
