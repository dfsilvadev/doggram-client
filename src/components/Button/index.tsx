import { forwardRef, ForwardRefRenderFunction } from "react";

import * as S from "./styles";

import { ButtonProps, WrapperStyleProps } from "./types";

const Button: ForwardRefRenderFunction<WrapperStyleProps, ButtonProps> = (
  {
    children,
    size = "medium",
    fullWidth = false,
    ghost = false,
    minimal = false,
    loading = false,
    icon,
    ...props
  },
  ref
) => {
  return (
    <S.Wrapper
      size={size}
      fullWidth={fullWidth}
      hasIcon={!!icon}
      minimal={minimal}
      ghost={ghost}
      ref={ref}
      {...props}
    >
      {loading ? (
        <S.Loader />
      ) : (
        <>
          {!!icon && icon}
          {!!children && <span>{children}</span>}
        </>
      )}
    </S.Wrapper>
  );
};

export default forwardRef(Button);
