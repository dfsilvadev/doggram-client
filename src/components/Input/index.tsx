import { ForwardRefRenderFunction, forwardRef } from "react";

import * as S from "./styles";

import { TextFieldProps } from "./types";

const Input: ForwardRefRenderFunction<HTMLInputElement, TextFieldProps> = (
  {
    label,
    icon,
    name,
    iconPosition = "left",
    disabled = false,
    error,
    ...props
  },
  ref
) => {
  return (
    <S.Wrapper disabled={disabled} error={!!error}>
      {!!label && <S.Label htmlFor={name}>{label}</S.Label>}
      <S.InputWrapper>
        {!!icon && <S.Icon iconPosition={iconPosition}>{icon}</S.Icon>}
        <S.Input
          type="text"
          disabled={disabled}
          name={name}
          {...(label ? { id: name } : {})}
          {...props}
          ref={ref}
        />
      </S.InputWrapper>
      {!!error && <S.ErrorMsg>{error}</S.ErrorMsg>}
    </S.Wrapper>
  );
};

export default forwardRef(Input);
