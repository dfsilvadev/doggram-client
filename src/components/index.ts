export { default as Avatar } from "./Avatar";
export { default as Box } from "./Box";
export { default as Button } from "./Button";
export { default as Card } from "./Card";
export { default as Container } from "./Container";
export { default as Form } from "./Form";
export { FlexComponent as Flex } from "./Flex";
export { default as Input } from "./Input";
export { default as Navbar } from "./Navbar";
export { default as SignInForm } from "./SignInForm";
export { default as SignUpForm } from "./SignUpForm";
