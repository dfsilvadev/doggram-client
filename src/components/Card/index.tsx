import { useLayoutEffect, useRef, useState } from "react";
import { ChatCentered, Heart } from "phosphor-react";

import { Flex } from "@/components";

import * as S from "./styles";

type CardProps = {
  imageUrl: string;
  likes?: number;
  comments?: number;
};

const Card = ({ imageUrl, likes = 0, comments = 0 }: CardProps) => {
  const [height, setHeight] = useState(366);
  const itemRef = useRef<HTMLDivElement>(null);

  useLayoutEffect(() => {
    const getGridWidth = () => {
      if (itemRef.current) {
        const gridColumnWidth = itemRef.current.offsetWidth;
        setHeight(gridColumnWidth);
      }
    };

    getGridWidth();
    window.addEventListener("resize", getGridWidth);

    return () => window.removeEventListener("resize", getGridWidth);
  }, []);

  return (
    <S.Item ref={itemRef} height={height}>
      <S.ItemImage src={imageUrl} />
      <S.Overlay>
        <Flex gap="1rem">
          <Flex gap="0.4rem">
            <Heart size={18} weight="bold" /> {likes}
          </Flex>

          <Flex gap="0.4rem">
            <ChatCentered size={18} weight="bold" /> {comments}
          </Flex>
        </Flex>
      </S.Overlay>
    </S.Item>
  );
};

export default Card;
