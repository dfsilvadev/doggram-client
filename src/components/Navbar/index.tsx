import { Link, NavLink } from "react-router-dom";
import { MagnifyingGlass, PlusCircle } from "phosphor-react";

import { Avatar, Button, Input } from "@/components";

import { ROUTES } from "@/utils/common/constant/routes";

import * as S from "./styles";

import logo from "@assets/images/svg/logo.svg";

const Navbar = () => {
  return (
    <S.Header>
      <S.Content>
        <S.Logo>
          <Link to={ROUTES.HOME}>
            <img
              src={logo}
              alt="logotipo petsgram: composto por um desenho vetor de um cachorro sorrindo e com uma mancha roxa com formato de triângulo em seu olho direito."
              title="logotipo petsgram: composto por um desenho vetor de um cachorro sorrindo e com uma mancha roxa com formato de triângulo em seu olho direito."
              aria-label="logotipo petsgram"
            />
          </Link>
        </S.Logo>

        <Input
          type="search"
          name="search"
          placeholder="search"
          icon={<MagnifyingGlass />}
          iconPosition="right"
          role="search"
          aria-label="search photo"
        />

        <S.Nav>
          <NavLink to={ROUTES.HOME}>
            <MagnifyingGlass size={24} weight="bold" aria-label="home link" />
          </NavLink>

          <Button
            icon={<PlusCircle size={24} weight="bold" />}
            minimal
            size="medium"
          />

          <Avatar src="https://avatars.githubusercontent.com/u/16245261?v=4" />
        </S.Nav>
      </S.Content>
    </S.Header>
  );
};

export default Navbar;
