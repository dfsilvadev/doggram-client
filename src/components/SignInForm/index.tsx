import { Link } from "react-router-dom";
import { EnvelopeSimple, Lock } from "phosphor-react";

import { Button, Form, Input } from "@/components";

import { ROUTES } from "@/utils/common/constant/routes";

import * as S from "./styles";

const SignInForm = () => {
  return (
    <Form aria-label="sign in form">
      <S.SignInFormGroup aria-label="form group">
        <Input name="email" placeholder="E-mail" icon={<EnvelopeSimple />} />
        <Input name="password" placeholder="Senha" icon={<Lock />} />

        <Button type="submit" fullWidth>
          Entrar
        </Button>

        <p>
          Não possui conta?
          <Link to={ROUTES.SIGNUP}>Faça seu cadastro agora!</Link>
        </p>
      </S.SignInFormGroup>
    </Form>
  );
};

export default SignInForm;
