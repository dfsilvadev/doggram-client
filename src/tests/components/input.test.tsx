import { act, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderWithTheme } from "@/utils/tests";
import { EnvelopeSimple } from "phosphor-react";

import { Input } from "@/components";

describe("Input", () => {
  it("should renders with label", () => {
    renderWithTheme(<Input label="Label" name="Field" />);

    expect(screen.getByLabelText("Label")).toBeInTheDocument();
  });

  it("should renders without label", () => {
    renderWithTheme(<Input />);

    expect(screen.queryByLabelText("Label")).not.toBeInTheDocument();
  });

  it("should renders with placeholder", () => {
    renderWithTheme(<Input placeholder="hey you" />);

    expect(screen.getByPlaceholderText("hey you")).toBeInTheDocument();
  });

  it("should changes its value when typing", async () => {
    renderWithTheme(<Input label="Label" name="Label" />);

    const input = screen.getByRole("textbox");
    const text = "This is my new text";

    act(() => {
      userEvent.type(input, text);
    });

    await waitFor(() => {
      expect(input).toHaveValue(text);
    });
  });

  it("should is accessible by tab", async () => {
    renderWithTheme(<Input label="Label" name="Label" />);

    expect(document.body).toHaveFocus();

    userEvent.tab();

    expect(await screen.findByLabelText("Label")).toHaveFocus();
  });

  it("should render an icon", () => {
    renderWithTheme(<Input icon={<EnvelopeSimple data-testid="icon" />} />);

    expect(screen.getByTestId("icon")).toBeInTheDocument();
  });

  it("should render an iconon the right side", () => {
    renderWithTheme(
      <Input
        icon={<EnvelopeSimple data-testid="icon" />}
        iconPosition="right"
      />
    );

    expect(screen.getByTestId("icon").parentElement).toHaveStyle({
      order: 1
    });
  });

  it("should does not changes its value when disabled", async () => {
    const onInput = jest.fn();
    renderWithTheme(
      <Input onInput={onInput} label="Input" name="Input" disabled />
    );

    const input = screen.getByRole("textbox");
    expect(input).toBeDisabled();

    const text = "This is my new text";

    act(() => {
      userEvent.type(input, text);
    });

    await waitFor(() => {
      expect(input).not.toHaveValue(text);
    });
    expect(onInput).not.toHaveBeenCalled();
  });

  it("should is not accessible by tab when disabled", () => {
    renderWithTheme(<Input label="Input" name="Input" disabled />);

    const input = screen.getByLabelText("Input");
    expect(document.body).toHaveFocus();

    act(() => {
      userEvent.tab();
    });

    expect(input).not.toHaveFocus();
  });

  it("should renders with error", () => {
    const { container } = renderWithTheme(
      <Input
        icon={<EnvelopeSimple data-testid="icon" />}
        label="Input"
        name="Input"
        error="Error message"
      />
    );

    expect(screen.getByText("Error message")).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});
