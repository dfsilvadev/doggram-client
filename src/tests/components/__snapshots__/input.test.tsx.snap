// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should renders with error 1`] = `
.c0 .c3 {
  border-color: #F16165;
  color: #9164fa;
}

.c0 .c5,
.c0 .c1 {
  color: #F16165;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.5rem;
  background: #f7f8fa;
  border-radius: 0.5rem;
  padding: 0 0.5rem;
  border: 2px solid transparent;
  -webkit-transition: 0.1s ease-in-out;
  transition: 0.1s ease-in-out;
}

.c4:focus-within {
  border: 2px solid #9164fa;
}

.c7 {
  color: #494d4b;
  font-family: Roboto,-apple-system,BlinkMacSystemFont,'Segoe UI',Poppins,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1rem;
  padding: 0.75rem 0;
  background: transparent;
  border: 2px solid transparent;
  outline: none;
  width: 100%;
}

.c2 {
  font-size: 0.875rem;
  color: #494d4b;
  font-weight: 500;
  cursor: pointer;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  color: #494d4b;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
}

.c6 > svg {
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  font-size: 1.4rem;
  line-height: 0;
}

.c3:focus-within .c5 {
  color: #9164fa;
}

.c8 {
  color: #F16165;
  font-size: 0.75rem;
  font-weight: 500;
  padding-top: 0.1rem;
}

<div
  class="c0"
>
  <label
    class="c1 c2"
    for="Input"
  >
    Input
  </label>
  <div
    class="c3 c4"
  >
    <div
      class="c5 c6"
    >
      <svg
        data-testid="icon"
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <path
          d="M32,56H224a0,0,0,0,1,0,0V192a8,8,0,0,1-8,8H40a8,8,0,0,1-8-8V56A0,0,0,0,1,32,56Z"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
        <polyline
          fill="none"
          points="224 56 128 144 32 56"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
      </svg>
    </div>
    <input
      class="c7"
      id="Input"
      name="Input"
      type="text"
    />
  </div>
  <p
    class="c8"
  >
    Error message
  </p>
</div>
`;
